building-and-packaging/material/systempaths-and-librarytools_slides.md

```diff
@@ -0,0 +1,289 @@
+
+- [Official homepage of FHS](https://refspecs.linuxfoundation.org/fhs.shtml)
+- Certain structure should
+
+**TODO** Set this link to the GitHub main branch before merging.
+
+---
+
+## Some Environment Variables
+
+- `HOME` or `~`: User's home directory
+- `PWD`: Current working directory
+- `PATH`: Location of executable
+- `LC_<NAME>`: Locale's
+- `LANG`: Systems' language
+- System variables often set from `/etc/profile` and `/etc/bashrc` or `/etc/bash.bashrc`
+
+---
+
+## Manipulating Environment Variables 1/2
+
+- Show content of paths via
+
+  ```bash
+  echo $VARIABLENAME
+  printenv VARIABLENAME
+  ```
+
+- Setting variable for a single command (bash)
+
+  ```bash
+  ENVVARIABLE=VALUE ./COMMAND
+  ```
+
+- Setting **environment** variable
+
+  ```bash
+  export ENVVARIABLE=VALUE
+  ```
+
+---
+
+## Manipulating Environment Variables 2/2
+
+- Adding additional values by colon (`:`) separated list
+
+  ```bash
+  PATH=ADDITONALPATH:${PATH}
+  ```
+
+  ```bash
+  PATH=/home/jaustar/bin:${PATH}
+  ```
+
+- Extra paths of user often set in `${HOME}/.bashrc`, `${HOME}/.bash_exports`, `${HOME}/.bash_profile`,...
+
+---
+
+## Some Environment Variables (C++)
+
+- `LD_<NAME>`: Loader related variable
+    - `LD_LIBRARY_PATH`: Library paths for runtime
+- `LIBRARY_PATH`: Library path for compilation (linker)
+- `CPATH`, `C_INCLUDE_PATH`, `CPLUS_INCLUDE_PATH`: Include file path
+- Names can vary beween OS, software.
+- Variables should be manipulated carefully.
+
+---
+
+## Paths and Environment Variables (Python)
+
+- `/usr/`: System-wide installation
+    - `/usr/bin`: Executable scripts
+    - `/usr/lib/pythonX.Y`: Modules/packages
+- `${HOME}/.local`: As above
+    - `${HOME}/.local/bin` and `${HOME}/.local/lib/pythonX.Y`
+- `PYTHONPATH`: Environment variable for non-standard installation of modules/packages
+    - Points to `prefix/lib/pythonX.Y/site-packages`
+
+---
+
+## Working with Libraries
+
+- How to find libraries?
+    - Paths?
+- How to incorporate libraries when compiling?
+    - Compileflags?
+
+---
+
+## Tools for Finding Libraries and Packages
+
+- [ldconfig](https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html)
+    - Managing paths of library loader
+- [pkg-config]((https://www.freedesktop.org/wiki/Software/pkg-config/))
+    - Manage package information such as
+        - Version
+        - Metadata
+        - Compile and link flags
+
+
+---
+
+## ldconfig
+
+- Generate links and cache to libraries
+- Can be used to inform system about new libraries at non-standard location
+    - Useful for your own codes/environments
+    - Usually automatically run by system's package manager (`apt`, `yum` etc.)
+- `LD_LIBRARY_PATH` commonly used, but only recommended for debugging
```

Comment:
I don't get this last point. Commonly used by users or by package managers? Who does the debugging?

Reply:
In some sense it used by both parties a lot. I am also not sure if there is a better way than manipulating `LD_LIBRARY_PATH`. On HPC platforms, where one uses tools like a module system, one does not want/cannot (root rights) call `ldconfig` the whole time. I will rephrase that and say something about this in the lecture.